File: clients/kmp/shared/src/commonTest/kotlin/org/wyrdsekai/app/network/ZoneBankSyncTest.kt
package org.wyrdsekai.app.network

import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

/** ZoneBankSync tests ( — KMP parity with RN). */
class ZoneBankSyncTest {

    private fun zone(id: String, user: String = "operator", added: Long = 1000, used: Long? = null) =
        ZoneBankEntry(zoneId = id, displayName = id, relayUrls = emptyList(), username = user, addedAt = added, lastUsedAt = used)

    /** A stub client backed by an in-memory server blob. */
    private class StubClient(var blob: String?, var stamp: Long = 0, val failGet: Boolean = false) : ZoneBankSyncClient {
        override suspend fun getZoneBank(): ZoneBankFetch? =
            if (failGet) null else ZoneBankFetch(blob, stamp)
        override suspend fun putZoneBank(bankJson: String, updatedAt: Long): Boolean {
            blob = bankJson; stamp = updatedAt; return true
        }
    }

    @Test
    fun merge_addsRemoteOnly() {
        val bank = ZoneBank()
        bank.addOrUpdateZone(zone("home-server"))
        val n = ZoneBankSync.mergeRemoteZones(bank, listOf(zone("relay-b")))
        assertEquals(1, n)
        assertEquals(setOf("home-server", "relay-b"), bank.zones.map { it.zoneId }.toSet())
    }

    @Test
    fun merge_keepsLocallyNewer() {
        val bank = ZoneBank()
        bank.addOrUpdateZone(zone("home-server", user = "local", used = 5000))
        val n = ZoneBankSync.mergeRemoteZones(bank, listOf(zone("home-server", user = "remote", used = 1000)))
        assertEquals(0, n)
        assertEquals("local", bank.getZone("home-server")!!.username)
    }

    @Test
    fun merge_takesRemotelyNewer() {
        val bank = ZoneBank()
        bank.addOrUpdateZone(zone("home-server", user = "local", used = 1000))
        val n = ZoneBankSync.mergeRemoteZones(bank, listOf(zone("home-server", user = "remote", used = 9000)))
        assertEquals(1, n)
        assertEquals("remote", bank.getZone("home-server")!!.username)
    }

    @Test
    fun sync_pullsMergesAndPushesUnion() = runTest {
        val bank = ZoneBank()
        bank.addOrUpdateZone(zone("home-server"))
        val remote = ZoneBank().also { it.addOrUpdateZone(zone("relay-b")) }
        val client = StubClient(blob = remote.serializeZones())
        val res = ZoneBankSync.syncZoneBank(bank, client, now = 7777)
        assertTrue(res.ok); assertEquals(1, res.pulled); assertTrue(res.pushed)
        assertEquals(setOf("home-server", "relay-b"), bank.zones.map { it.zoneId }.toSet())
        assertEquals(7777, client.stamp)
        // Server received the merged superset.
        val pushed = ZoneBank().also { it.load(null, client.blob) }
        assertEquals(setOf("home-server", "relay-b"), pushed.zones.map { it.zoneId }.toSet())
    }

    @Test
    fun sync_pushesLocalWhenServerEmpty() = runTest {
        val bank = ZoneBank()
        bank.addOrUpdateZone(zone("home-server"))
        val client = StubClient(blob = null)
        val res = ZoneBankSync.syncZoneBank(bank, client, now = 100)
        assertTrue(res.ok); assertEquals(0, res.pulled); assertTrue(res.pushed)
        assertEquals(1, ZoneBank().also { it.load(null, client.blob) }.zones.size)
    }

    @Test
    fun sync_noPushWhenBothEmpty() = runTest {
        val client = StubClient(blob = null)
        val res = ZoneBankSync.syncZoneBank(ZoneBank(), client, now = 100)
        assertTrue(res.ok); assertEquals(0, res.pulled); assertFalse(res.pushed)
        assertNull(client.blob)
    }

    @Test
    fun sync_reportsGetFailure() = runTest {
        val res = ZoneBankSync.syncZoneBank(ZoneBank(), StubClient(blob = null, failGet = true), now = 1)
        assertFalse(res.ok)
        assertEquals("zonebank get failed", res.error)
    }
}

File: clients/rn/src/server/__tests__/zoneBankSync.test.ts
/**
 * zoneBankSync tests ( cross-device sync).
 * Mock secureStorage; drive the real zoneBankStore; stub the sync client.
 */
const mem = new Map<string, string>();
jest.mock('../../state/secureStorage', () => ({
  secureStorage: {
    async getItem(k: string) { return mem.has(k) ? mem.get(k)! : null; },
    async setItem(k: string, v: string) { mem.set(k, v); },
    async removeItem(k: string) { mem.delete(k); },
  },
}));

import {
  mergeRemoteZones,
  serializeBank,
  syncZoneBank,
  type ZoneBankSyncClient,
} from '../zoneBankSync';
import { useZoneBankStore, type ZoneBankEntry } from '../../state/zoneBankStore';
import type { NatsResult } from '../NatsServerClient';

const HOME_SERVER: ZoneBankEntry = {
  zoneId: 'home-server', displayName: 'home-server', relayUrls: ['wss://relay-node:4443'],
  username: 'operator', addedAt: 1000, lastUsedAt: 1000,
};
const QF: ZoneBankEntry = {
  zoneId: 'relay-b', displayName: 'example-relay', relayUrls: ['wss://qf:4443'],
  username: 'operator', addedAt: 2000,
};

beforeEach(() => {
  mem.clear();
  useZoneBankStore.setState({ relays: [], zones: [], loaded: true });
});

/** A stub client backed by an in-memory server blob. */
function makeClient(initialBlob: string | null, updatedAt = 0): {
  client: ZoneBankSyncClient;
  stored: () => { bank: string | null; updatedAt: number };
} {
  let blob = initialBlob;
  let stamp = updatedAt;
  const client: ZoneBankSyncClient = {
    async getZoneBank(): Promise<NatsResult<{ bank: string | null; updatedAt: number }>> {
      return { ok: true, data: { bank: blob, updatedAt: stamp } };
    },
    async putZoneBank(bankJson, ts): Promise<NatsResult<{ updatedAt: number }>> {
      blob = bankJson;
      stamp = ts;
      return { ok: true, data: { updatedAt: ts } };
    },
  };
  return { client, stored: () => ({ bank: blob, updatedAt: stamp }) };
}

describe('mergeRemoteZones', () => {
  it('adds remote-only zones', () => {
    useZoneBankStore.setState({ relays: [], zones: [HOME_SERVER], loaded: true });
    const changed = mergeRemoteZones([QF]);
    expect(changed).toBe(1);
    expect(useZoneBankStore.getState().zones.map((z) => z.zoneId).sort()).toEqual(['home-server', 'relay-b']);
  });

  it('keeps the locally-newer entry (LWW)', () => {
    const localNewer = { ...HOME_SERVER, username: 'local-name', lastUsedAt: 5000 };
    useZoneBankStore.setState({ relays: [], zones: [localNewer], loaded: true });
    const remoteOlder = { ...HOME_SERVER, username: 'remote-name', lastUsedAt: 1000 };
    const changed = mergeRemoteZones([remoteOlder]);
    expect(changed).toBe(0);
    expect(useZoneBankStore.getState().getZone('home-server')?.username).toBe('local-name');
  });

  it('takes the remotely-newer entry (LWW)', () => {
    const localOlder = { ...HOME_SERVER, username: 'local-name', lastUsedAt: 1000 };
    useZoneBankStore.setState({ relays: [], zones: [localOlder], loaded: true });
    const remoteNewer = { ...HOME_SERVER, username: 'remote-name', lastUsedAt: 9000 };
    const changed = mergeRemoteZones([remoteNewer]);
    expect(changed).toBe(1);
    expect(useZoneBankStore.getState().getZone('home-server')?.username).toBe('remote-name');
  });

  it('ignores malformed entries', () => {
    const changed = mergeRemoteZones([{ zoneId: '' } as ZoneBankEntry, undefined as unknown as ZoneBankEntry]);
    expect(changed).toBe(0);
    expect(useZoneBankStore.getState().zones).toHaveLength(0);
  });
});

describe('syncZoneBank', () => {
  it('pulls remote zones, merges, and pushes the union', async () => {
    useZoneBankStore.setState({ relays: [], zones: [HOME_SERVER], loaded: true });
    const { client, stored } = makeClient(JSON.stringify([QF]));
    const res = await syncZoneBank(client, 7777);
    expect(res).toEqual({ ok: true, pulled: 1, pushed: true });
    // Local now has both.
    expect(useZoneBankStore.getState().zones.map((z) => z.zoneId).sort()).toEqual(['home-server', 'relay-b']);
    // Server received the merged superset with our stamp.
    const up = stored();
    expect(up.updatedAt).toBe(7777);
    const pushed = JSON.parse(up.bank!) as ZoneBankEntry[];
    expect(pushed.map((z) => z.zoneId).sort()).toEqual(['home-server', 'relay-b']);
  });

  it('pushes local zones when the server is empty', async () => {
    useZoneBankStore.setState({ relays: [], zones: [HOME_SERVER], loaded: true });
    const { client, stored } = makeClient(null);
    const res = await syncZoneBank(client, 100);
    expect(res.ok).toBe(true);
    expect(res.pulled).toBe(0);
    expect(res.pushed).toBe(true);
    expect(JSON.parse(stored().bank!)).toHaveLength(1);
  });

  it('does not push when both local and server are empty', async () => {
    const { client, stored } = makeClient(null);
    const res = await syncZoneBank(client, 100);
    expect(res).toEqual({ ok: true, pulled: 0, pushed: false });
    expect(stored().bank).toBeNull();
  });

  it('reports a get failure without throwing', async () => {
    const client: ZoneBankSyncClient = {
      async getZoneBank() { return { ok: false, error: 'boom' }; },
      async putZoneBank() { return { ok: false, error: 'unused' }; },
    };
    const res = await syncZoneBank(client, 1);
    expect(res.ok).toBe(false);
    expect(res.error).toBe('boom');
  });

  it('serializeBank excludes nothing but zones (no secrets present anyway)', () => {
    useZoneBankStore.setState({ relays: [], zones: [HOME_SERVER, QF], loaded: true });
    const blob = JSON.parse(serializeBank()) as ZoneBankEntry[];
    expect(blob).toHaveLength(2);
    // Sanity: no password field exists on the entry shape.
    expect(Object.keys(blob[0])).not.toContain('password');
  });
});
